rfd/0001-general-design.md

```diff
@@ -0,0 +1,288 @@
+││A│      │ │┌─────────────┐│    │┌─────────────┐│    │┌─────────────┐│ │    │A│  │
+││T├──────┼─▶│   Active    ││    ││   Active    ││    ││   Active    │◀─┼────┤T│  │
+││O│      │ │└─────────────┘│    │└─────────────┘│    │└─────────────┘│ │    │O│  │
+││M│      │ └───────────────┘    └───────────────┘    └───────────────┘ │    │M│  │
+││I│      │                                                             │    │I│  │
+││C│      └─────────────────────────────▲───────────────────────────────┘    │C│  │
+│└┬┘                              ┌─────┴──────┐                             └┬┘  │
+│ │  ┌──────────────┐             │Mark !Active│          ┌────────────────┐  │   │
+│ │  │              │             │ Append new │          │                │  │   │
+│ └──│   Strategy   │             └─────┬──────┘       ┌─▶│ Health-watcher │──┘   │
+│    │              │              ┌────┴─────┐        │  │                │      │
+│    └──────────────┴────┐         │  Update  │        │  └────────────────┘      │
+│                      ╔═╩════╦────▶   Mutex  │        │      ┌─────────┐         │
+│                      ║ Each ║    └────▲─────┘        │      │   Add   │         │
+│                      ║Next()║         │              └──────┤  Route  │         │
+│                      ╚══════╝         │                     └─────────┘         │
+└───────────────────────────────────────┼─────────────────────────────────────────┘
+                                   ┌────┴────┐
+                                   │ Update  │
+                                   │ Routes  │
+                                   └─────────┘ 
+```
+Forwarder provides the reference to the slice of `[]*Route`.
+
+#### Property updates
+Both `Strategy` and `Health-Watcher` can use the slice and use `atomics` to update
+each `*Route` individually.
+
+#### Pool updates
+In case if `[]*Route` should change Forwarder will block on the `Update Mutex` adding to the
+slice and marking entities that should be removed as inactive (as they still can be in reference
+with the strategy which will block only on Next() entry).
+
+Updates are rare, this design does not target additional optimizations.
+
+#### Strategy Next()
+Strategy will use the current reference for provide the selection mechanics as a forward iteration loop.
+
+Strategy will block `updateMutex` on `Next()` but in following way:
+```go
+func (s Strategy) Next() *Route {
+	s.fwd.updateMutex.Lock()
+	s.fwd.updateMutex.Unlock()
+	// ... rest of the code
+	// range s.fwd.Routes
+}
+```
+This will adjust behavior to not pick the stale `[]*Route` if Update Routes started
+to mark the entries and replacing the slice reference.
+
+However, it will try to use the stale list once if it was fetched before Update Routes happened.
+
+### Rate Limiter
+
+###### Per-session: When TLS Handshake happened and session got authorization
+In this path we provide a Token bucket for each customer pool resource, which will
```

Comment:
Is this limiting generally or per client identity ?

Reply:
As the configuration was not shared explicitly, it seems to be an open question; however, the following strategies can be applicable:
- per client pool (mentioned, as well normal behavior, as usually price is per LB )
- per client overall

In the case of per client account overall, it probably would be a Token Bucket with atomics attached to the customer map[clientUUID]*RateLimiter. However trying to imagine the case when client creates multiple forwarders and still accounted for the account and not per service :)